// ngether/lib/modules/ngetherMod.introduceMe.js

var understandIntroduceMe = {
	id: 'understandIntroduceMe',
	executor: function(next, thisEscalator) {
		thisEscalator.thisNgether.listenTo(
			[ 'ngether/introduceMe/' ],
			[ thisEscalator.thisNgether.config.id ],
			function(socketMessage, receivedTopic, incomingSocket) {
				if (typeof incomingSocket.ngether !== 'object') return false;
				var alreadyConnected = thisEscalator.thisNgether.isConnectedTo(socketMessage.id);
				if (alreadyConnected !== false && socketMessage.id !== thisEscalator.thisNgether.id) {
					thisEscalator.thisNgether.log.add('DELETING DUPLICATE CONNECTION: ' + alreadyConnected, 'error', 'understandIntroduceMe');
					delete thisEscalator.thisNgether.connections[alreadyConnected];
				};
				incomingSocket.ngether.partnerConfig = {
					type: socketMessage.type,
					id: socketMessage.id,
					host: socketMessage.host,
					port: socketMessage.port,
					poolIds: socketMessage.poolIds
				};
				thisEscalator.thisNgether.log.add('was introduced to ' + socketMessage.id, 'plain', 'understandIntroduceMe');
				return true;
			}
		);
		thisEscalator.thisNgether.log.add('understanding topic introduceMe !', 'job', 'init');
		next();
	}
};


module.exports = understandIntroduceMe;
