// ngether/lib/ngetherSpot.js
var fs = require('fs');
var classes = {
	_logger: require('./logger'),
	_bubPubSub: require('bubpubsub'),
	_sugar: require('./sugar'),
	_mongodb: require('mongodb'),
	_funcs: require('./funcs'),
	_engineIO: require('engine.io'),
	_engineIOclient: require('engine.io-client'),
};
var funcs = new classes._funcs();
var ngether = require('./ngether.js');


// -------------------- ngether SPOT OBJECT
var ngetherSpot = function(config, callback) {
	var thisNgether = this;
	Object.merge(this, ngether.prototype, true);
	this.config = this.defaultConfig();
	Object.merge(this.config, config, true);
	this.id = this.config.id;
	this.connections = {};
	this.subscriptions = {};
	this.commandHandlers = {};	
	this.meshBubPubSub = new classes._bubPubSub('mesh');
	this.meshBubPubSub.defaults.debugging = 1;
	this.meshBubPubSub.defaults.throwErrors = true;
	this.meshBubPubSub.subscribe(
		'/toMesh/',
		this._bubPubSubBridge_ToMesh,
		{ getBubbles: true, scope: thisNgether },
		'bubPubSubBridge_ToMesh'
	);
	this.initBase(
		function() {
			thisNgether.initSpot(
				function() {
					// DEBUG
					console.log('SPOT SPOT SPOT DONE.....');
					if (typeof callback === 'function') callback(thisNgether);
				}
			);
		}
	);
	return true;
};


ngetherSpot.prototype.defaultConfig = function() {
	var defaults = {
		id: 'ngetherSpot_' + funcs.makeId(12),
		hubAddress: {
			host: 'localhost',
			port: 9999
		},
		poolIds: [],
		allowDirectNetworking: true,
		listConnections: false
	};
	return defaults;
};


// --------------------- init spot
ngether.prototype.initSpot = function(callback) {
	var mySpot = this;
	var modSteps = mySpot.modules;
	var initSteps = [];
	var step04 = {
		id: 'repl',
		executor: function(next, thisEscalator) {
			thisEscalator.thisNgether = mySpot;
			funcs.openSocktREPL(mySpot.id, mySpot.config.replPort);
			mySpot.log.add('REPL live on port: ' + mySpot.config.replPort, 'job', 'init');
			next();
		}
	};
	initSteps.push(step04);
	var step06 = {
		id: 'socketServer',
		executor: function(next, thisEscalator) {
			mySpot.startSocketServer(mySpot.config.socketServer.port);
			mySpot.log.add('engine.IO server listening on port ' + mySpot.config.socketServer.port, 'job', 'init');
			next();
		}
	};
	initSteps.push(step06);
	var understandPubSubs = {
		id: 'understandPubSubs',
		executor: function(next, thisEscalator) {
			mySpot._bubPubSubBridge_FromMesh(next);
		}
	};
	initSteps.push(understandPubSubs);
	var step07 = {
		id: 'connectToHubs',
		executor: function(next, thisEscalator) {
			for (var n in mySpot.config.hubs) {
				mySpot.connectTo(mySpot.config.hubs[n]);
			}
			next();
		}
	};
	initSteps.push(step07);
	for (var i in modSteps) initSteps.push(modSteps[i]);
	var step15 = {
		id: 'initComplete',
		executor: function(next, thisEscalator) {
			mySpot.log.add('[ngetherSpot] module loaded.');
			if (typeof callback === 'function') callback(mySpot);
			next();
		}
	};
	initSteps.push(step15);
	var init = new funcs.escalator('ngetherSpotInit', initSteps, false).start();
};


module.exports = ngetherSpot;
console.log('[ngetherSpot] module loaded.');
